com.unity.ml-agents/Runtime/SideChannels/SideChannel.cs

```diff
@@ -28,16 +30,125 @@ public abstract class SideChannel
         /// Is called by the communicator every time a message is received from Python by the SideChannel.
         /// Can be called multiple times per simulation step if multiple messages were sent.
         /// </summary>
-        /// <param name="data"> the payload of the message.</param>
-        public abstract void OnMessageReceived(byte[] data);
+        /// <param name="msg">The incoming message.</param>
+        public abstract void OnMessageReceived(IncomingMessage msg);
 
         /// <summary>
         /// Queues a message to be sent to Python during the next simulation step.
         /// </summary>
         /// <param name="data"> The byte array of data to be sent to Python.</param>
-        protected void QueueMessageToSend(byte[] data)
+        protected void QueueMessageToSend(OutgoingMessage msg)
         {
-            MessageQueue.Add(data);
+            MessageQueue.Add(msg.ToByteArray());
+        }
+    }
+
+    public class IncomingMessage : IDisposable
+    {
+        byte[] m_Data;
+        Stream m_Stream;
+        BinaryReader m_Reader;
+
+        public IncomingMessage(byte[] data)
+        {
+            m_Data = data;
+            m_Stream = new MemoryStream(data);
+            m_Reader = new BinaryReader(m_Stream);
+        }
+
+        public int ReadInt32()
+        {
+            return m_Reader.ReadInt32();
+        }
+
+        public float ReadFloat32()
+        {
+            return m_Reader.ReadSingle();
+        }
+
+        public string ReadString()
+        {
+            var strLength = ReadInt32();
+            var str = Encoding.ASCII.GetString(m_Reader.ReadBytes(strLength));
+            return str;
+        }
+
+        public IList<float> ReadFloatArray()
```

Comment:
Why not a float[]?
Not sure we need this method to be honest.

Reply:
I added it on the python side in response to @awjuliani's comment on the doc.

`float[]` implements the `IList<float>` interface; I figured it was better to be a bit more generic in case we ever wanted to change the underlying implementation.